### ISSUE TITLE
pull request

### ISSUE THREAD

**Issue by edgarriba** [edgarriba/OpenDroneMap#11]
@dakotabenjamin @smathermather @paulinus

The baby starts to walk.
Let's test everything in order to do a proper PR in the master branch.

We can also discuss here the project design in order to be able to move things around before doing this huge merge.

**Comment by dakotabenjamin** [comment]
What kind of parallelization is happening, specifically with opensfm? These processes seem to be taking longer than normal, even on a large (32-core) machine. I'm trying to do some comprehensive testing but right now run time is slowing me down. 

It took 45 minutes to run odm_data_bellus, vs 18 minutes using the master

**Comment by edgarriba** [comment]
AFAIK parallelization happens only during the features detection and matching process. Not sure if Ceres has parallel computing for BA task? @paulinus

**Comment by dakotabenjamin** [comment]
Can you help me get the python testing started? My unfamiliarity with ecto really has me struggling to set up the tests. I think I understand conceptually that we need to initialize an ecto plasm, then we can call each cell and run tests on each- however I am having trouble getting that basis. You can see here what I mean: https://github.com/dakotabenjamin/OpenDroneMap/blob/nose_tests/tests/test_odm.py

**Comment by edgarriba** [comment]
sure! Firstly, in this case what do you want to test? Secondly, what I would do is take one of the datasets that produces a proper result and use it as a ground truth reference. Finally, design a test for each cell using as inputs/outputs the ground truth files. We should define very well every test before start coding.

**Comment by dakotabenjamin** [comment]
Ok So I think this is a start of what should be tested in python. I will use a subset of the toledo dataset. 

## opendm/

### config.py
- test each parameter (min, max, out of bounds, etc.)

### context.py
- Each path must be a valid, existing path

### io.py
- check each function with a known string

### system.py
- (Get_ccd_widths is depreciated)
- parse_coordinate_system:
  - This is also old code - new one is in types

### types.py
- ODM_Photos
  - updated focal equal to known focal length
  - updated ccd equal to known ccd
- ODM_Georef
  - calculate_epsg needs to be updated, see master
  - convert_to_las: assert existence of las file output
  - utm_to_latlon:
    - 
    - 

## scripts/

### dataset.py
- test that supported_extensions works with a variety of file names, even bogus ones
- check outputs are all the photos

### resize.py
- resulting images are the right size 
  - `img.shape[0] or img.shape[1] == params.resize_to`
- metadata has been properly updated
  - `img.metadata['Exif.Photo.PixelX/YDimension'] == params.resize_to`

### OpenSfM.py
- config file is contains info same as params
- when matcher_distance > 0 it is written to the config
- at least one reconstruction file is generated
- check that bundler file is exported (possible to check if valid?)

### CMVS.py
- validate params
- system.run() command is equal to some known string

### PMVS.py
- validate params
- system.run() command is equal to some known string

### odm_meshing.py
- validate params
- system.run() command is equal to some known string

### odm_texturing.py
- validate params
- system.run() command is equal to some known string

### odm_georeferencing.py
- validate params
- system.run() command is equal to some known string when using EXIF coords
- system.run() when using GCP

### odm_orthophoto.py
- validate params
- system.run() command is equal to some known string

**Comment by edgarriba** [comment]
sounds good to me! We could use this unit test framework:
https://cloud.google.com/appengine/docs/python/tools/localunittesting?csw=1

**Comment by dakotabenjamin** [comment]
Interesting. I had been looking at using [Travis-CI](https://travis-ci.com/) because of the strong support for open-source projects, and it integrates with Github well. What do you think google app engine can offer that would better suite this project?

**Comment by edgarriba** [comment]
well, I'm not confident at all with Travis since I've never used. But we can give it a shot! Regarding gtest I have some experience with the c++ framework (used by opencv) but not in python. I only know that have some well defined macros for testing.

**Comment by dakotabenjamin** [comment]
I finally got a testsing framework in place. I made a branch on the main repo from the cmake branch for the code sprint and I've pushed the testing code to as 'tests' folder there. 

BTW, I had to use the `git merge` option `-s recursive -Xtheirs` to basically completely override your code into the branch.

**Comment by edgarriba** [comment]
yes, just seen that. Thanks to merge it. I realized that the sprint starts tomorrow. I'm a bit busy but I will help as much as possible.

**Comment by smathermather** [comment]
Sounds great, @edgarriba.
